flang/lib/Optimizer/Analysis/AliasAnalysis.cpp

```diff
@@ -30,8 +30,15 @@ using namespace mlir;
 
 static bool isDummyArgument(mlir::Value v) {
   auto blockArg{mlir::dyn_cast<mlir::BlockArgument>(v)};
-  if (!blockArg)
+  if (!blockArg) {
+    auto defOp = v.getDefiningOp();
+    if (defOp) {
+      if (auto declareOp = mlir::dyn_cast<fir::DeclareOp>(defOp))
+        if (declareOp.getDummyScope())
+          return true;
```

Comment:
Slightly modifying the sample from the test so that `%arg0` is now a local allocation:
```
func.func @test1(%arg1: !fir.ref<f32> {fir.bindc_name = "y", fir.target}) {
  %alloca = fir.alloca i32
  %scope_out = fir.dummy_scope : !fir.dscope
  %0 = fir.declare %alloca {fortran_attrs = #fir.var_attrs<target>, uniq_name = "_QFtestEx"} : (!fir.ref<f32>, !fir.dscope) -> !fir.ref<f32>
  %1 = fir.declare %arg1 dummy_scope %scope_out {fortran_attrs = #fir.var_attrs<target>, uniq_name = "_QFtestEy"} : (!fir.ref<f32>, !fir.dscope) -> !fir.ref<f32>
  %2 = fir.load %1 : !fir.ref<f32>
  fir.store %2 to %0 : !fir.ref<f32>
  %scope_in1 = fir.dummy_scope : !fir.dscope
  %3 = fir.declare %0 dummy_scope %scope_in1 {uniq_name = "_QFtestFinnerEx"} : (!fir.ref<f32>, !fir.dscope) -> !fir.ref<f32>
  %4 = fir.declare %1 dummy_scope %scope_in1 {uniq_name = "_QFtestFinnerEy"} : (!fir.ref<f32>, !fir.dscope) -> !fir.ref<f32>
  %5 = fir.load %4 : !fir.ref<f32>
  fir.store %5 to %3 : !fir.ref<f32>
  %scope_in2 = fir.dummy_scope : !fir.dscope
  %6 = fir.declare %0 dummy_scope %scope_in2 {uniq_name = "_QFtestFinnerEx"} : (!fir.ref<f32>, !fir.dscope) -> !fir.ref<f32>
  %7 = fir.declare %1 dummy_scope %scope_in2 {uniq_name = "_QFtestFinnerEy"} : (!fir.ref<f32>, !fir.dscope) -> !fir.ref<f32>
  %8 = fir.load %7 : !fir.ref<f32>
  fir.store %8 to %6 : !fir.ref<f32>
  return
}
```

`%3` and `%0` should alias. But I think the current logic will determine that `%3` is a dummy argument due to its alias scope. Dummy arguments do not alias with local allocations and so we would incorrectly assume that they do not alias.

Reply:
Ahh hang on. `getInstantiationPoint` will be `false` and so `getSource` will track through the `fir.declare` in my example.

`getSource` does use `isDummyArgument` in other places though, and I am worried about a similar confusion. For example, if an unknown operation leads to an unknown source, then the above mechanism could occur (I think maybe this could happen when using mlir inlining inside of an openmp or openacc container operation).

Reply:
The usage of `isDummyArgument` at line [307](https://github.com/llvm/llvm-project/pull/92472/files#diff-5d0a764398edd1ffe8b17f16f7be874b9178ca3a90f3a7457eec76d7403457f8R307) allows propagating the source search past the indirection (the load).  My change does allow propagating past `fir.declare` (when `getInstantiationPoint` is `false`) in this case, but I cannot figure out an example where this would produce incorrect results in the end.

The `isDummyArgument` usage at the end of `getSource` should also not make a difference when `getInstantionPoint` is `false`, because we will just pass through any `fir.declare` we meet on the path.

Reply:
I am worried about [L372](https://github.com/llvm/llvm-project/pull/92472/files#diff-5d0a764398edd1ffe8b17f16f7be874b9178ca3a90f3a7457eec76d7403457f8L372).

If for some reason we have an unknown source (e.g. due to a block argument or some unsupported operation), we will return that it is an argument if `isDummyArgument` returns true. That would then follow the logic I outlined above in `fir::AliasAnalysis::alias` leading to strong assumptions that it doesn't alias with other things in the same function.

Maybe `isDummyArgument` needs something like `getInstantiationPoint` to control whether inlined dummy arguments are treated as dummy arguments. Alternatively `fir::AliasAnalysis::alias` needs to understand dummy scopes.

Reply:
Thanks for your explanation and for baring with me while I figure out that this is correct. I think this is good as it is.